voicevox_engine/app/routers/speaker.py

```diff
@@ -151,6 +175,61 @@ def singer_info(speaker_uuid: str, core_version: str | None = None) -> SpeakerIn
             speaker_uuid=speaker_uuid,
             speaker_or_singer="singer",
             core_version=core_version,
+            resource_baseurl=resource_baseurl,
+            resource_format=resource_format,
+        )
+
+    # リソースはAPIとしてアクセスするものではないことを表明するためOpenAPIスキーマーから除外する
+    @router.get(f"/{RESOURCE_ENDPOINT}/{{resource_name}}", include_in_schema=False)
+    def resources(
+        resource_name: str,
+        if_none_match: Annotated[str | None, Header()] = None,
+        if_modified_since: Annotated[str | None, Header()] = None,
+    ) -> Response:
+        """
+        ResourceManagerから発行されたURLへのアクセスに対応する
+        """
+        resource_path = resource_manager.resource_path(resource_name)
+        if resource_path is None or not resource_path.exists():
+            raise HTTPException(status_code=404)
+        response = FileResponse(
+            resource_path,
+            headers={
+                "Cache-Control": "max-age=2592000, immutable, stale-while-revalidate=2592000"
+            },
+            stat_result=resource_path.stat(),
+        )
+        res_headers = response.headers
+        # 304レスポンスの作成
+        modified_headers = {
+            k: res_headers.get(k)
+            for k in [
+                "Cache-Control",
+                "Content-Location",
+                "Date",
+                "ETag",
+                "Expires",
+                "Vary",
+            ]
+        }
+        modified_response = Response(
+            status_code=304,
+            headers={k: v for k, v in modified_headers.items() if v is not None},
         )
+        # ETagとLast-Modifiedの検証
+        if if_none_match is not None:
+            etag = res_headers["ETag"]
+            if etag in [tag.strip(" W/") for tag in if_none_match.split(",")]:
+                return modified_response
+        elif if_modified_since is not None:
+            _if_modified_since = parsedate(if_modified_since)
+            last_modified = parsedate(res_headers["Last-Modified"])
+            if (
+                _if_modified_since is not None
+                and last_modified is not None
+                and _if_modified_since >= last_modified
+            ):
+                return modified_response
```

Comment:
レスポンスの検証を実装
ヘッダのパース部分はstarletteを確認したが正確ではない気が…
https://github.com/encode/starlette/blob/0.37.2/starlette/staticfiles.py#L220-L234

https://developer.mozilla.org/ja/docs/Web/HTTP/Caching#%E6%A4%9C%E8%A8%BC
https://developer.mozilla.org/ja/docs/Web/HTTP/Headers/If-None-Match
https://developer.mozilla.org/ja/docs/Web/HTTP/Headers/If-Modified-Since
https://developer.mozilla.org/ja/docs/Web/HTTP/Status/304